Implement search faiss index & code refactor

adaorachi opened the pull request:
### Changes Made:

- Search Functionality with Faiss Index:
   - Implemented search functionality using Faiss index to efficiently search through embeddings.
   - Added functions to send response data to client

- Code Refactoring:
   - Refactored existing codebase to improve readability and maintainability.
   - Modularized code into separate functions and classes for better organization.
   - Introduced a configuration file (config.json) to store application settings and parameters.
   - Centralized configuration management for easier maintenance and customization
   
- Type Definitions and Annotations:
   - Added type definitions and annotations throughout the codebase for improved code clarity and error prevention.
   - Ensured type safety by specifying data types for function parameters, return values, and variables.

adaorachi committed:
resolve isort linter

adaorachi committed:
implement search functionality

adaorachi commented on backend/multimodalexplorer/functions/fetch_embed.py:
```diff
@@ -0,0 +1,73 @@
+
+    try:
+        # Load raw data
+        raw_data = pd.read_csv(file_path_raw, sep="\t")
+
+        # Load UMAP embeddings
+        with open(file_path_embed, "rb") as file:
+            embeddings = np.load(file)
+
+        # Merge raw data with UMAP embeddings
+        merged_data = pd.concat([raw_data, pd.DataFrame(embeddings)], axis=1)
+
+        # Rename columns
+        merged_data.rename(
+            columns={
+                merged_data.columns[-3]: "x",
+                merged_data.columns[-2]: "y",
+                merged_data.columns[-1]: "cluster",
+            },
+            inplace=True,
+        )
+
+        # Convert merged data to list of dictionary records
+        merged_embed_data = merged_data.to_dict("records")
```
Here, I'm currently merging the 'umap_embedding' NumPy array with the raw data in CSV format within a dictionary. This combined data is what I'm sending to the client to fetch the embeddings. The purpose behind this approach is to gather all the embedding data upfront and have it ready to draw on the canvas when the visualizer is zoomed in on the user interface (UI). This helps avoid making repeated calls to the UI.

Initially, I tried making another call to the UI when the visualizer is sufficiently zoomed (as requested in the planned tasks), but this resulted in multiple calls being rendered as the user interacts with the window. I'm unsure if the current implementation is the most efficient way to handle this.

Mortimerp9 replied:
I am not sure I understand. What do you call raw data? Is this the original text before embedding? Where is this raw_data_file created?

adaorachi replied:
Yes, this is the initial text prior to embedding. The logic for saving raw_data is contained within the _save_data function in the process_dataset script.
This file is generated and the raw_data is appended to it concurrently while I encode and save the embeddings in batches and chunks. It's formatted as a TSV file with two columns: data and media_type.

Mortimerp9 commented on .pre-commit-config.yaml:
```diff
@@ -1,38 +0,0 @@
-default_language_version:
```
why are you removing this one?

adaorachi replied:
A YAML file already exist in the backend folder. During the debugging of the linter issue, I moved it in the root folder. 
So, I removed it afterwards because I wanted to be sure it is okay to move it to the root folder permanently. Additionally, having it in the root folder streamlines linter checks during pre-commit.

Mortimerp9 commented on backend/README.md:
````diff
@@ -8,6 +8,71 @@ Add dependencies with `poetry add`
+   cd MultiModalExplorer/backend/multimodalexplorer`
+   ```
+
+3. Create a virtual environment (optional but recommended):
+
+   ```bash
+   python -m venv venv
+   ```
+
+4. Activate the virtual environment (skip this step if you didn't create a virtual environment):
+
+   On Windows:
+
+   ```bash
+   venv\Scripts\activate
+   ```
+
+   On macOS and Linux:
+
+   ```bash
+   source venv/bin/activate
+   ```
````
poetry should do all of this for you:

https://python-poetry.org/docs/basic-usage/#using-your-virtual-environment

Mortimerp9 commented on backend/multimodalexplorer/api/endpoints/search.py:
```diff
@@ -0,0 +1,42 @@
+        data = await request.json()
+
+        if "search_data" not in data or "search_type" not in data:
+            return {"error": "Missing required fields 'search_data' or 'search_type'"}
```
you should raise an error: https://fastapi.tiangolo.com/tutorial/handling-errors/

Mortimerp9 commented on backend/multimodalexplorer/api/endpoints/search.py:
```diff
@@ -0,0 +1,42 @@
+
+            return JSONResponse(content={"data": search_result}, status_code=200)
+        except Exception as e:
+            return {"error": f"Failed to search index: {str(e)}"}
```
same as above, raise the error, and also, log the exception so that it's visible in the server logs.

Mortimerp9 commented on backend/multimodalexplorer/api/endpoints/search.py:
```diff
@@ -0,0 +1,42 @@
+    @router.post("/search_data")
+    async def get_search_result(request: Request) -> Dict[str, Any]:
+
+        data = await request.json()
```
the benefit of fastapi is that it will deal with param typing and checks for you, see https://fastapi.tiangolo.com/tutorial/body/#import-pydantics-basemodel

so you can just declare a "model" for your data and let fastapi check that the schema is respected by the query

Mortimerp9 commented on backend/multimodalexplorer/api/endpoints/search.py:
```diff
@@ -0,0 +1,42 @@
+
+        data = await request.json()
+
+        if "search_data" not in data or "search_type" not in data:
+            return {"error": "Missing required fields 'search_data' or 'search_type'"}
+
+        search_query = {
+            "search_data": data["search_data"],
+            "search_type": data["search_type"],
+            "search_src_lang": data.get("search_src_lang", "eng_Latn"),
+        }
```
using fastapi+pydantic would:
1. document this expectation
2. deal with this whole code for you

Mortimerp9 commented on backend/multimodalexplorer/config.json:
```diff
@@ -12,9 +12,10 @@
   "umap_file": { "dir": "artifact/umap", "ext": "npy" },
```
could you write in the readme a quick doc about this file?

Mortimerp9 commented on backend/multimodalexplorer/functions/fetch_embed.py:
```diff
@@ -0,0 +1,73 @@
+        logger.info(f"Loaded UMAP embeddings with shape: {embeddings.shape}")
+
+        return merged_embed_data
+    except FileNotFoundError:
+        logger.error(f"UMAP embeddings file not found at: {file_path_embed}")
+        return None
+    except Exception as e:
+        logger.exception(f"An error occurred while loading UMAP embeddings: {e}")
+        return None
```
you call this in the fastapi endpoint, with a try catch, but you never raise any exception from here. You should either reraise here or just let the original exception bubble up there and deal with it.

Mortimerp9 commented on backend/multimodalexplorer/functions/reduce_embed_dims.py:
```diff
@@ -41,44 +43,67 @@ def __init__(
-            logger.exception(f"An error occurred while loading UMAP embeddings: {e}")
-            return None
+        # Perform clustering
+        clusters = kmeans.fit_predict(embeddings)
```
https://umap-learn.readthedocs.io/en/latest/clustering.html suggests that density clustering is better suited for umapped data. We might get some better results with HDBSCAN as per their example code.

Mortimerp9 commented on backend/multimodalexplorer/functions/search_faiss_index.py:
```diff
@@ -87,7 +88,31 @@ def _process_search_query(self, search_query: dict) -> np.ndarray:
+        dir_path, ext = self.raw_data_file.values()
+        file_path = get_file_path(dir_path, ext)
+
+        raw_data = pd.read_csv(file_path, sep="\t")
```
this means that you are reading the whole data at every query, which is going to be very slow. You should maintain this data in memory or at least in a memory mapped structure. pyarrow should help you with that: https://arrow.apache.org/docs/python/csv.html You shouldn't even have to drop to pandas, just use `to_table` and use pyarrow table for quick access: https://arrow.apache.org/docs/python/data.html#tables

Mortimerp9 reviewed (changes requested):
looks good, I have a few changes that I think would be good to add.

adaorachi committed:
implement pyarrow and pydantic

adaorachi committed:
fix bug in get_file_path utils func

adaorachi committed:
fix isort/black linter

adaorachi commented:
@Mortimerp9 Thank you immensely for your comprehensive review and invaluable feedback. I've implemented the changes as per your suggestions.

### Implemented Changes:
- Implemented error handling of HTTP request errors where necessary.
- Employed pyarrow for efficient reading of the raw_data file with a memory-mapped structure.
- Adopted HDBSCAN - a density-based clustering algorithm for enhanced unmapped data clustering.
- Provided a concise documentation regarding the config.json - configuration specifications in the README file.
- Utilized pydantic models for enhancing route parameter typing and validation.

Kindly review these modifications at your earliest convenience and share any further insights or suggestions you may have.

artemru commented on backend/multimodalexplorer/functions/process_dataset.py:
```diff
@@ -88,7 +90,8 @@ def _extract_data_from_batch(self, batch: Dict[str, Any]) -> List[Any]:
             if "set" in batch:
-                return [e[0] for e in batch["set"]]
+                data = [e[0] for e in batch["set"]]
+                return [s for s in data if isinstance(s, str) and len(s) <= 514]
```
why this constant=514 ? could we truncate this string to this number ?
